File: avalon-modules/avalon-dataset/avalon-dataset-service/src/main/java/io/github/lishangbu/avalon/dataset/service/impl/TypeServiceImpl.java
package io.github.lishangbu.avalon.dataset.service.impl;

import io.github.lishangbu.avalon.dataset.entity.Type;
import io.github.lishangbu.avalon.dataset.entity.Type_;
import io.github.lishangbu.avalon.dataset.repository.TypeRepository;
import io.github.lishangbu.avalon.dataset.service.TypeService;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.ExampleMatcher;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/// 属性服务实现。
@Service
@RequiredArgsConstructor
public class TypeServiceImpl implements TypeService {
    private final TypeRepository typeRepository;

    @Override
    public Page<Type> getPageByCondition(Type type, Pageable pageable) {
        return typeRepository.findAll(
                Example.of(
                        type,
                        ExampleMatcher.matching()
                                .withIgnoreNullValues()
                                .withMatcher(
                                        Type_.NAME,
                                        ExampleMatcher.GenericPropertyMatchers.contains())
                                .withMatcher(
                                        Type_.INTERNAL_NAME,
                                        ExampleMatcher.GenericPropertyMatchers.contains())),
                pageable);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public Type save(Type type) {
        return typeRepository.save(type);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public Type update(Type type) {
        return typeRepository.save(type);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public void removeById(Long id) {
        typeRepository.deleteById(id);
    }

    @Override
    public List<Type> listByCondition(Type type) {
        ExampleMatcher matcher =
                ExampleMatcher.matching()
                        .withIgnoreNullValues()
                        .withMatcher(Type_.NAME, ExampleMatcher.GenericPropertyMatchers.contains())
                        .withMatcher(
                                Type_.INTERNAL_NAME,
                                ExampleMatcher.GenericPropertyMatchers.contains());
        return typeRepository.findAll(Example.of(type, matcher));
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-service/src/main/kotlin/io/github/lishangbu/avalon/dataset/service/impl/TypeServiceImpl.kt
package io.github.lishangbu.avalon.dataset.service.impl

import io.github.lishangbu.avalon.dataset.entity.dto.SaveTypeInput
import io.github.lishangbu.avalon.dataset.entity.dto.TypeSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.TypeView
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateTypeInput
import io.github.lishangbu.avalon.dataset.repository.TypeRepository
import io.github.lishangbu.avalon.dataset.service.TypeService
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/** 属性服务实现*/
@Service
class TypeServiceImpl(
    /** 属性仓储*/
    private val typeRepository: TypeRepository,
) : TypeService {
    /** 保存属性*/
    @Transactional(rollbackFor = [Exception::class])
    override fun save(command: SaveTypeInput): TypeView = TypeView(typeRepository.save(command.toEntity(), SaveMode.INSERT_ONLY))

    /** 更新属性*/
    @Transactional(rollbackFor = [Exception::class])
    override fun update(command: UpdateTypeInput): TypeView = TypeView(typeRepository.save(command.toEntity(), SaveMode.UPDATE_ONLY))

    /** 按 ID 删除属性*/
    @Transactional(rollbackFor = [Exception::class])
    override fun removeById(id: Long) {
        typeRepository.deleteById(id)
    }

    /** 按条件查询属性列表*/
    override fun listByCondition(specification: TypeSpecification): List<TypeView> = typeRepository.listViews(specification)
}
